### PR TITLE
buildroot: Make it easier to rebuild all of the images after an upgrade/change

### PR DESCRIPTION
**Pull request by nathanchance**
Commit messages should be fairly self explanatory, let me know if there are any questions/concerns!

### CODE REVIEW

**Comment by MSF-Jarvis** [comment]
I was trying to run this for the buildroot upgrade PR earlier and never quite figured out how, this is a welcome change :hugs:. Trying it out right now, will submit review once done.

**Comment by MSF-Jarvis** [comment]
@nathanchance I can't figure out why my build fails, anything obvious I may have missed? https://del.dog/raw/qiqopafori

**Comment by nickdesaulniers** [comment]
> make[1]: Leaving directory '/home/msfjarvis/git-repos/continuous-integration/buildroot/src/output/build/host-gcc-initial-8.3.0/build'
> output/images/rootfs.cpio could not be found! Did the build error?

**Comment by MSF-Jarvis** [comment]
> make[1]: Leaving directory '/home/msfjarvis/git-repos/continuous-integration/buildroot/src/output/build/host-gcc-initial-8.3.0/build'
> output/images/rootfs.cpio could not be found! Did the build error?

Yeah Nathan helped me find the problem over Telegram, for some reason my C++ compiler was there but configure refused to use it.

**Comment by nathanchance** [comment]
Thanks! @nickdesaulniers did you have any comments?

File: buildroot/rebuild.sh [hunk 1]
```diff
@@ -1,44 +1,62 @@
...
+        make menuconfig
+        make savedefconfig
     fi
-    cp -v "${IMAGE}" "${IMAGES_FOLDER}"
+
+    # Build images
+    make -j"$(nproc)"
+
+    # Make sure images folder exists
+    IMAGES_FOLDER=../../images/$(basename "${CONFIG//.config}")
```
Comment (nickdesaulniers) [inline buildroot/rebuild.sh]: was the double `//` intentional?

Comment (nathanchance) [reply buildroot/rebuild.sh]: Yes, it's a habit (`/` removes the first occurrence, `//` removes all occurrences). Here it obviously does not matter so I can change it if it helps make it clearer.

File: buildroot/rebuild.sh [hunk 2]
```diff
@@ -5,13 +5,21 @@
...
+    cd src || exit 1
+    if [[ $(git describe --exact-match --tags HEAD) != "${BUILDROOT_VERSION}" ]]; then
+        git fetch origin ${BUILDROOT_VERSION}
+        git checkout ${BUILDROOT_VERSION}
+    fi
+
+    # Clean up artifacts from the last build
+    make clean
+else
+    git clone -b ${BUILDROOT_VERSION} git://git.busybox.net/buildroot src
```
Comment (nickdesaulniers) [inline buildroot/rebuild.sh]: should be a shallow clone?

**Comment by nickdesaulniers** [comment]
> This will save bandwidth across image upgrades/rebuilds.

How's that?  Seems like a compressed tarball would be smaller than all of git history?

File: buildroot/rebuild.sh (hunk 2 above)
Comment (nathanchance) [reply buildroot/rebuild.sh]: Sure, I can make that change.

**Comment by nathanchance** [comment]
> > This will save bandwidth across image upgrades/rebuilds.
> 
> How's that? Seems like a compressed tarball would be smaller than all of git history?

I guess I was comparing it to redownloading the tarball over and over in the current scheme. I suppose what I could do is just not remove the downloaded tarball source every rebuild and use the `print-version` make target to check if we need to update the source. At the same time, shallow clone + new tags is a little bit easier yet should achieve the same thing.

File: buildroot/rebuild.sh (hunk 1 above)
Comment (nickdesaulniers) [reply buildroot/rebuild.sh]: sorry, what do you mean by "removes" ?

File: buildroot/rebuild.sh (hunk 2 above)
Comment (nickdesaulniers) [reply buildroot/rebuild.sh]: That may change the fetch case to:
```
$ git fetch && git reset --hard FETCH_HEAD
```

**Comment by nickdesaulniers** [comment]
> I guess I was comparing it to redownloading the tarball over and over in the current scheme.

Ah, I see.  In that case, I think it's simpler to wrap the fetch in a check if the file does not already exist, then fetch it.  Might be nice to check the checksum of it either way if buildroot provided them.

File: buildroot/rebuild.sh (hunk 1 above)
Comment (nathanchance) [reply buildroot/rebuild.sh]: Sorry, removes the part after the slashes from the variable's value (an example makes it clearer):

```
$ CONFIG=arm.config.config; echo ${CONFIG/.config}
arm.config
$ CONFIG=arm.config.config; echo ${CONFIG//.config}
arm
```

Comment (nickdesaulniers) [reply buildroot/rebuild.sh]: oh, maybe create a new variable on the line before use with a comment; I feel like I'll forget that's what `//` does in the future. Thanks for the explanation.

**Comment by nathanchance** [comment]
Alright, I think I have addressed all comments :)

File: buildroot/rebuild.sh [hunk 3]
```diff
@@ -21,15 +26,14 @@ done
 # Download latest buildroot release
 BUILDROOT_VERSION=2019.02.3
 if [[ -d src ]]; then
-    cd src || exit 1
-    if [[ $(git describe --exact-match --tags HEAD) != "${BUILDROOT_VERSION}" ]]; then
-        git fetch origin ${BUILDROOT_VERSION}
-        git checkout ${BUILDROOT_VERSION}
+    if [[ $(cd src && make print-version | cut -d - -f 1 >/dev/null) != "${BUILDROOT_VERSION}" ]]; then
```
Comment (nickdesaulniers) [inline buildroot/rebuild.sh]: if you redirect stdout from `cut` to `/dev/null`, is the subshell ever going to return a string to compare against `"${BUILDROOT_VERSION}"`?

File: buildroot/rebuild.sh [hunk 4]
```diff
@@ -3,7 +3,16 @@
 
 function download_br() {
     mkdir -p src
-    curl https://buildroot.org/downloads/buildroot-"${BUILDROOT_VERSION}".tar.gz | tar -xzf - -C src --strip-components=1
+    TARBALL=buildroot-${BUILDROOT_VERSION}.tar.gz
+    rm -rf "${TARBALL}"
```
Comment (nickdesaulniers) [inline buildroot/rebuild.sh]: probably don't need `-r` for a tarball

File: buildroot/rebuild.sh [hunk 5]
```diff
@@ -3,7 +3,16 @@
...
+    curl -LO https://buildroot.org/downloads/"${TARBALL}"
+    SHA256=31a0dd7f5df75d20f6b74f0684918b643e5bf08d87dd5f9f02525852eecccf95
```
Comment (nickdesaulniers) [inline buildroot/rebuild.sh]: add it as a file, then use `sha256sum -c <file>`
see: https://shapeshed.com/unix-sha1sum/

File: buildroot/rebuild.sh (hunk 3 above)
Comment (nathanchance) [reply buildroot/rebuild.sh]: Ah whoops, this should be `2>/dev/null` :(

File: buildroot/rebuild.sh (hunk 4 above)
Comment (nathanchance) [reply buildroot/rebuild.sh]: Yes, true.

File: buildroot/rebuild.sh (hunk 5 above)
Comment (nathanchance) [reply buildroot/rebuild.sh]: Sure

**Comment by nathanchance** [comment]
Thanks for the review all!

**Pull request merged**